src/parser.rs

```diff
@@ -0,0 +1,365 @@
+            let consequent = self.get_operator(tokenlist)?;
+            match implicator {
+                Implicator(direction) => match direction {
+                    Direction::UniDirectional => Ok(!antecedent | consequent),
+                    Direction::BiDirectional => Ok(antecedent == consequent),
+                },
+                _ => Err(anyhow!("No implicator found")),
+            }
+        } else {
+            Err(anyhow!("Unexpected end of token list"))
+        }
+    }
+
+    fn get_operator<I>(&mut self, tokenlist: &mut Peekable<I>) -> Result<bool>
+    where
+        I: Iterator<Item = &'a Token>,
+    {
+        let mut node = self.get_factor(tokenlist);
+        while let Some(Operator(op)) = tokenlist.peek() {
+            node = match tokenlist.next() {
+                Some(Operator('+')) => Ok(node? & self.get_factor(tokenlist)?),
+                Some(Operator('|')) => Ok(node? | self.get_factor(tokenlist)?),
+                Some(Operator('^')) => Ok(node? ^ self.get_factor(tokenlist)?),
+                _ => Err(anyhow!("Found unexpected operator: {:?}", op)),
+            }
+        }
+        node
+    }
+
+    fn get_factor<I>(&mut self, tokenlist: &mut Peekable<I>) -> Result<bool>
+    where
+        I: Iterator<Item = &'a Token>,
+    {
+        match tokenlist.next() {
+            Some(Parenthesis('(')) => {
+                let res = self.get_operator(tokenlist);
+                match tokenlist.next() {
+                    Some(Parenthesis(')')) => res,
+                    _ => Err(anyhow!("Missing closing parenthesis")),
+                }
+            }
+            Some(Operator('!')) => Ok(!self.get_factor(tokenlist)?),
+            Some(Bool(b)) => Ok(*b),
+            _ => Err(anyhow!("Unexpected end of token list")),
+        }
+    }
+
+    pub fn evaluate(&mut self, input: &str) -> Result<bool> {
+        let tokenlist = self.tokenize(input).context("Could not tokenize input")?;
+        self.get_rule(&mut tokenlist.iter().peekable())
+            .context("Syntactical error")
+    }
+}
+
+// PermutationList is an iterator that iterates over all permutations of a rule input string
+pub struct PermutationList<'a> {
+    formula: &'a str,
+    pub variables: Vec<char>,
+    size: usize,
+}
+
+impl PermutationList<'_> {
+    pub fn new(formula: &str) -> PermutationList {
+        let mut set = HashSet::new();
+        let mut variables = formula
+            .chars()
+            .filter_map(|c| match c {
+                'A'..='Z' if set.insert(c) => Some(c),
+                _ => None,
+            })
+            .collect::<Vec<char>>();
+        variables.sort_unstable();
+        PermutationList {
+            formula,
+            variables,
+            size: 0,
+        }
+    }
+}
+
+impl Iterator for PermutationList<'_> {
+    type Item = String;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.size == 1 << self.variables.len() {
+            None
```

Comment:
Should `size` not also be rest to allow for repeated iterations?

Reply:
You mean reset right? AFAIK in iterators' purpose is just to get consumed and disregarded afterward. For our usecase this is all we need because we need just one PermutationList iterator per rule to create 1 unique truthtable per rule so there's no need to reuse it.

https://doc.rust-lang.org/stable/std/iter/#implementing-iterator

Reply:
Ye I did mean reset and that makes. `PermutationIter` seems like a better name to me then